src/StringTools/InternableString.cs

```diff
@@ -0,0 +1,343 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.StringTools
+{
+    /// <summary>
+    /// Represents a string that can be converted to System.String with interning, i.e. by returning an existing string if it has been seen before
+    /// and is still tracked in the intern table.
+    /// </summary>
+    /// <remarks>
+    /// The structure is public because it's exposed via the <see cref="TryInternStringDelegate"/> callback. It's a mutable struct and should
+    /// generally not be used directly from outside of the library.
+    /// </remarks>
+    public ref struct InternableString
```

Comment:
`IInternable` was able to represent any string like structure whereas `InternableString` only works with spans. The assumption is that any string like structure that msbuild wants to intern should be convertible to a span, which as far as I can tell is a pretty good assumption.